### PR TITLE
Webpack 4.0

### PR DESCRIPTION
**Pull request by gauravtiwari**
This PR adds webpack 4.0 support to Webpacker. 

### TODO: 

 - [x] Use new webpack 4 development and production modes
 - [x] Upgrade webpack dev server
 - [x] Pre-release of WebpackManifestPLugin compatible plugin
 - [x] Mini css extract plugin
 - [x] Vue loader
 - [x] Use new built-in options
 - [ ] Update readme
 - [x] Test examples

### Blockers
- https://github.com/danethurber/webpack-manifest-plugin/pull/128
- HMR support in mini-css-extract-plugin
- Vue loader [#1196](https://github.com/vuejs/vue-loader/issues/1196)

### CODE REVIEW

**Comment by renchap** [comment]
It might be a good idea to replace `extract-text-plugin` with https://github.com/webpack-contrib/mini-css-extract-plugin once HMR support is merged.

**Comment by gauravtiwari** [comment]
Thanks @renchap that's the plan 👍

**Comment by montogeek** [comment]
What is the status of this PR? Is there something we can help from webpack side?

**Comment by rossta** [comment]
@montogeek Note the blocker in the description. This project relies on plugin that hasn't merged v4 support yet.

**Comment by montogeek** [comment]
@rossta Already commented on it

File: package/environments/production.js [hunk 1]
```diff
@@ -11,7 +10,7 @@ module.exports = class extends Base {
 
     this.plugins.append(
       'UglifyJs',
-      new UglifyJsPlugin({
+      new webpack.optimize.UglifyJsPlugin({
```
Comment (montogeek) [inline package/environments/production.js]: This won't work with ES6 modules

**Comment by gauravtiwari** [comment]
This PR is going to take a while since the plugin and loader support isn't 100% yet - the main blockers are manifest and vue loader, everything else seems to be working fine. 

Will release a beta once this PR is merged and then we can incrementally enhance in separate PRs until stable 4.0.x

**Comment by apuntovanini** [comment]
@gauravtiwari any hints on how to get it working / upgrading?
Thanks!

**Comment by gauravtiwari** [comment]
@apuntovanini Pretty simple and no breaking changes (AFAIK) - just update gem and npm package as above ^^

**Comment by saneef** [comment]
@apuntovanini You might have to run `bundle exec rails webpacker:install` if you are upgrading on the existing project.

**Comment by apuntovanini** [comment]
:) yep, thats the reason I asked, mine is broken, saying no manifest found. I use CommonsChunkManifest (now deprecated, isn't it?), but even without it doesnt work

**Comment by apuntovanini** [comment]
ah, I see, didn't try that, I will try again and let you know ASAP, thanks!

**Comment by gauravtiwari** [comment]
@apuntovanini Oh yes, that's true. Please see this gist on how to migrate common chunk - https://gist.github.com/sokra/1522d586b8e5c0f5072d7565c2bee693#configurate-cache-groups

**Comment by gauravtiwari** [comment]
With webpacker it would be,

```js
environment.config.merge({
  optimization: {
    splitChunks: {
      cacheGroups: {
        commons: {
          test: /[\\/]node_modules[\\/],
          name: "vendors",
          chunks: "all"
        }
      }
    }
  }
})

**Comment by apuntovanini** [comment]
Wonderful, it does work (`bundle exec rails webpacker:install` was indeed needed)! Now running webpack 4 with no issues! Thanks a lot!

**Comment by vcavallo** [comment]
I see `webpack: ^3.11.0` in my yarn.lock - is that correct? I was

**Comment by rposborne** [comment]
It looks like one blocker with https://github.com/danethurber/webpack-manifest-plugin/pull/134 has been solved.

**Comment by danielpowell4** [comment]
@gauravtiwari Any timeline for moving 4.x to stable?

If there is a clear checklist/roadmap I'm sure there are a good few of us who will help out!

**Comment by gauravtiwari** [comment]
@danielpowell4 Thanks, basically it's nearly there. I have made this PR (https://github.com/rails/webpacker/pull/1546) a while ago but couldn't finish it, apologies :(

Will try to get this done over the weekend and see if I can publish a stable release next week.

**Comment by danielpowell4** [comment]
@gauravtiwari no need to apologize! Timeline sounds awesome.

I have been on the pre version since it came out. Only recently ran into a snag when I went to tweak my babel config.

If you need help writing docs or monitoring the repo, feel free to reach out!

**Comment by iChip** [comment]
What's involved in getting 4.x released? Happy to assist.

**Comment by gauravtiwari** [comment]
Hey, Sorry, I need to find sometime to review some PRs like: #1694 and go through this big list of issues: https://github.com/rails/webpacker/issues to see if we have missed anything obvious that should be in 4.x stable release.

What if I release a beta 4.x version? Would that help time being?

**Comment by gauravtiwari** [comment]
Just made another pre-release:

```
yarn add @rails/webpacker@next
bundle update webpacker
```

**Comment by gauravtiwari** [comment]
Also, now Rails 6.0-alpha ships with Webpacker as default- https://github.com/rails/rails

**Pull request merged**